### DISCUSSION TITLE
How to use different Bluetooth midi

### DISCUSSION THREAD

**Discussion by lcollins1989** [Builty/TonexOneController#8] [General]
Hello!

Ok I'm learning a lot and I promise I am researching too. I just thought I'd ask the question.

I've been digging through the midi control task to understand how the Bluetooth midi works. I bought a wireless midi adapter (Xvive MD1) and I'm trying to figure out what needs changing in the code to make this work. From what I've discovered the Xvive does use the standard BLE midi protocol.

I couldn't figure out how to use serial midi with the recommended waveshare board as the RX pin doesn't have a breakout.

**Comment by Builty** [comment]
Hey.
Yeah Bluetooth  is a bit complicated. There is Bluetooth servers, and Bluetooth clients.
A client is a device that listens out for a peripheral device, and then connects to it, and receives commands.
A server is a device that advertises its precense, and allows a client to connect and receives commands.

A device like the M-Vave Chocolate footswitch is a server device. My code version V1.0.1.2 has a client implementation, that listens for the advertising from the M-Vave, and connects to it.

Now yesterday I finished implementing a Server mode. Code is committed to main branch but I haven't released builds for it. Another user is helping to test it.
In the latest source you can select if you want Server or Client.

In Server mode, the controller advertises its precense, and other devices (like a phone) can see it, connect and then send Midi commands. 
I posted a video of this to Youtube a few days ago, where a phone is connected, and a Midi app on the phone allows preset switching.

Now looking at the Xvive MD1, its a server device. So it shouldn't need yesterdays big changes for Server mode.
What we need to know is the "device name" for the MD1. Refer to midi_control.c line 98:

// M-vave Chocolate device name is 'FootCtrl'
static const char remote_device_name[20] = {"FootCtrl"};

When scanning for devices, if this name is matched, then I connect to it and search for the standard Bluetooth Midi UUIDs.
So we need to know what the device name is of the MD1.
Then the code could be changed to replace the Chocolate's name with the MD1 name.
Ideally it would probably be a list of device names, with menuconfig used to enable/disable each one.

I'm happy to make the changes, just need to know what the device name is.
For me I would use the app "LightBlue" to scan, and it would reveal all of the details of the device.

As for Serial Midi, yeah I wasn't thinking of that when I picked the Waveshare board. It can still be done, but would need a mod wire connected to the PCB UART TX output, so it can bypass the RS485 transceiver. And that mod wire would connect to an external board with the optocoupler and connectors.

**Reply by lcollins1989** [reply]
![image](https://github.com/user-attachments/assets/8e59d08b-f532-4b83-9103-f3b23cdaa394)

Ok I found it with the light blue app and this is what came up!

**Reply by lcollins1989** [reply]
![image](https://github.com/user-attachments/assets/c24c048b-18c3-428d-9081-c9546b33024a)

**Comment by Builty** [comment]
Yep, so "FootCtrl" would need to replaced with "Xvive MD1" in the device name.
You can see the Midi UUIDs shown match with the ones in my source:
// Midi service UUID
// 03b80e5a-ede8-4b33-a751-6ce34ec4c700

// 7772e5db-3868-4112-a1a9-f2669d106bf3   Midi characteristic

So it should just work. Are you able to build the code? 
Later today I could work on adding menu config enables and code to make a list of names to scan for.

**Reply by lcollins1989** [reply]
Amazing! I will give this a try and get back to you.

**Comment by Builty** [comment]
Also, you would need to send Midi Program Change commands to switch presets on the Tonex One. With the program being a 0-based index.
I.e. 80 80 C0 00 would select the first preset on the Tonex.

**Reply by Builty** [reply]
I've committed a little change, that lets you select which Midi BT devices to scan for with menu config. 
Added the Xvive MD1 in there also.

**Reply by lcollins1989** [reply]
Oh wow, you work fast!

**Comment by Builty** [comment]
I got a message that you had some trouble building the source, but can't find it here.
Hopefully you got it sorted.

I could make a test build for you if you like

**Reply by lcollins1989** [reply]
I would love a test build if possible.

I realized that I was asking what was probably a really simple question about VSC so it was better in hindsight for me to research and figure it out myself.

**Reply by Builty** [reply]
Have a go with this one. Its Bluetooth Client with the Xvive enabled.
[TonexController_V1.0.2.1_test.zip](https://github.com/user-attachments/files/17708290/TonexController_V1.0.2.1_test.zip)

**Comment by eykh** [comment]
Hi,
as far as I know, it is unusual to limit the names of the devices that are allowed to connect. I have several CME WIDI devices. They connect to whatever is found "on-air"! For example they also connect to my NUX Mighty Air practice amp. And they also can be connected to phones and desktop operating systems. It's impossible for these devices to know the names of partner devices upfront. The list would obviously be endless.

This implies that the devices have to be turned on in a specific order and that those two have to be the only ones that are scanning/advertising at that time. But that is the design of BLE MIDI. 
CME introduced "groups". Even non CME devices could join such a group. The group's host/server stores the client's ID and only allows these to connect. The other devices are set to a client-only mode, so that they don't accidently connect to each other. When the group is established, all devices of that group can communicate wirelessly with each other!

Long story short: I'm quite sure, that the restriction to certain device names should be disabled. I will try it. I will also test the CME WIDI group compatibility.

**Reply by Builty** [reply]
I understand what you are saying. The scan should be filtering based on the Midi UUID.
I can work on this.

**Reply by eykh** [reply]
You are too fast! :) Let me do some tests with different wireless MIDI devices first. I think the two available options make the current users happy.

**Reply by eykh** [reply]
aah, okay, I got it! That MIDI service UUID is common to all BLE MIDI devices! Very good!
